streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -389,6 +391,18 @@ public void close() {
         metrics.removeAllSensors();
     }
 
+    void closeProducer() {
+        if (exactlyOnceEnabled) {
+            try {
+                recordCollector.close();
+            } catch (final Throwable e) {
+                log.error("{} Failed to close producer: ", logPrefix, e);
+            }
+        } else {
+            throw new IllegalStateException("Tasks should only close producers if exactly-once semantics is enabled.");
```

Comment:
Is this necessary? There is a guard here `if(exactlyOnceEnabled)` there is also the same in `StreamThread` line 422. Could we remove the one from `StreamThread` and just not throw the `IllegalStateException` ? What is it giving us?